demos/viscoelastic/viscoelastic_loading.py

```diff
@@ -0,0 +1,517 @@
+#
+# For starters let's use one of the default meshes provided by Firedrake, `RectangleMesh`. We have chosen 40 quadrilateral elements in the $x$ direction and 40 quadrilateral elements in the $y$ direction. It is worth emphasising that the setup has coarse grid resolution so that the demo is quick to run! For real simulations we can use fully unstructured meshes to accurately resolve important features in the model, for instance near coastlines or sharp discontinuities in mantle properties.  We can print out the grid resolution using `log`, a utility provided by G-ADOPT. (N.b. `log` is equivalent to python's `print` function, except that it simplifies outputs when running simulations in parallel.)
+#
+# On the mesh, we also denote that our geometry is Cartesian, i.e. gravity points
+# in the negative z-direction. This attribute is used by G-ADOPT specifically, not
+# Firedrake. By contrast, a non-Cartesian geometry is assumed to have gravity
+# pointing in the radially inward direction.
+#
+# Boundaries are automatically tagged by the built-in meshes supported by Firedrake. For the `RectangleMesh` being used here, tag 1 corresponds to the plane $x=0$; 2 to the $x=L$ plane; 3 to the $y=0$ plane; and 4 to the $y=D$ plane. For convenience, we can rename these to `left_id`, `right_id`, `bottom_id` and `top_id`.
+
+# +
+# Set up geometry:
+L = 1500e3  # length of the domain in m
+D = 2891e3  # Depth of domain in m
+nx = 40  # Number of horizontal cells
+nz = 40  # number of vertical cells
+
+# Let's print out the grid resolution in km
+log(f"Horizontal resolution {L/nx/1000} km")
+log(f"Vertical resolution {D/nz/1000} km")
+
+mesh = RectangleMesh(nx, nz, L, D, name="mesh", quadrilateral=True)
+mesh.cartesian = True
+
+# Shift top of mesh to z = 0.
+mesh.coordinates.dat.data[:, 1] -= D
+
+left_id, right_id, bottom_id, top_id = 1, 2, 3, 4  # Boundary IDs
+# -
+# We now need to choose finite element function spaces. `V` , `W`, `TP1` and `R` are symbolic
+# variables representing function spaces. They also contain the
+# function space’s computational implementation, recording the
+# association of degrees of freedom with the mesh and pointing to the
+# finite element basis. We will choose Q2-Q1 for the mixed incremental displacement-pressure similar to our mantle convection demos. This is a Taylor-Hood element pair which has good properties for Stokes modelling. We also initilaise a discontinuous tensor function space that wil store our previous values of the deviatoric stress, as the gradient of the continous incremental displacement field will be discontinuous.
+
+# Set up function spaces - currently using the bilinear Q2Q1 element pair:
+V = VectorFunctionSpace(mesh, "Q", 2)  # (Incremental) Displacement function space (vector)
+W = FunctionSpace(mesh, "Q", 1)  # Pressure function space (scalar)
+TP1 = TensorFunctionSpace(mesh, "DQ", 2)  # (Discontinuous) Stress tensor function space (tensor)
```

Comment:
Maybe me being a pedant, but it would be nice if functions spaces were all consistently named - perhaps 1 letter for each? S or Y maybe instead of TP1?